src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.hpp

```diff
@@ -59,52 +59,126 @@ class ShenandoahEvacOOMCounter {
  * Provides safe handling of out-of-memory situations during evacuation.
  *
  * When a Java thread encounters out-of-memory while evacuating an object in a
- * load-reference-barrier (i.e. it cannot copy the object to to-space), it does not
- * necessarily follow we can return immediately from the LRB (and store to from-space).
- *
- * In very basic case, on such failure we may wait until the evacuation is over,
- * and then resolve the forwarded copy, and to the store there. This is possible
- * because other threads might still have space in their GCLABs, and successfully
- * evacuate the object.
- *
- * But, there is a race due to non-atomic evac_in_progress transition. Consider
- * thread A is stuck waiting for the evacuation to be over -- it cannot leave with
- * from-space copy yet. Control thread drops evacuation_in_progress preparing for
- * next STW phase that has to recover from OOME. Thread B misses that update, and
- * successfully evacuates the object, does the write to to-copy. But, before
- * Thread B is able to install the fwdptr, thread A discovers evac_in_progress is
- * down, exits from here, reads the fwdptr, discovers old from-copy, and stores there.
- * Thread B then wakes up and installs to-copy. This breaks to-space invariant, and
- * silently corrupts the heap: we accepted two writes to separate copies of the object.
- *
- * The way it is solved here is to maintain a counter of threads inside the
- * 'evacuation path'. The 'evacuation path' is the part of evacuation that does the actual
- * allocation, copying and CASing of the copy object, and is protected by this
- * OOM-during-evac-handler. The handler allows multiple threads to enter and exit
- * evacuation path, but on OOME it requires all threads that experienced OOME to wait
- * for current threads to leave, and blocks other threads from entering. The counter state
- * is striped across multiple cache lines to reduce contention when many threads attempt
- * to enter or leave the protocol at the same time.
+ * load-reference-barrier (i.e. it cannot copy the object to to-space), a special
+ * protocol is required to assure that all threads see the same version of every
+ * object.
+ *
+ * This file and its accompanying .cpp and .inline.hpp files hold the implementation
+ * of this protocol.  The general idea is as follows:
+ *
+ *  1. If we fail to evacuate the entirety of live memory from all cset regions,
+ *     we will transition to STW full gc at the end of the evacuation cycle.  Full GC
```

Comment:
First it attempts a degenerated cycle. If we don't believe a degenerated oom-during-evac cycle can ever succeed, we should change this behavior.

Reply:
Is this true?  I believe degenerated cycle is unable to recover from OOM during evac.  We always have to upgrade to Full GC in this situation.  That's because the OOM during Evac protocol has compromised the cset by possibly allowing references to the cset to sneak through the LRB barrier.  Hereafter, degen is not sufficient to recover.

Reply:
Yes, this happens. That is why we specifically has to add this patch: https://github.com/openjdk/shenandoah/blob/master/src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp#L202 to help the degenerated cycle recover from oom-during-evac correctly.

Reply:
I am wrong about this.  So if the _degen_point is _degenerated_evac, the degenerated evac cycle sets the update watermark value on every active region to the current top (because we may need to rescan ranges of regions that are above the original update watermark (because we may have written a leaked cset pointer into this memory) and then we iterate through all cset regions to evacuate all objects not previously evacuated.  There is no attempt by degen gc to resume evacuation where concurrent GC left off.

Reply:
Yes, that's my read of it. What I missed earlier is that once any thread fails to evacuate, the other threads are prevented from evacuation. This increases the chances for a degenerated cycle to successfully conclude the evacuation because these other threads may still have memory available in their GCLABs.